sim-lib/src/random_activity.rs

```diff
@@ -0,0 +1,211 @@
+use core::fmt;
+use std::collections::HashMap;
+use std::fmt::Display;
+
+use bitcoin::secp256k1::PublicKey;
+use rand_distr::{Distribution, Exp, LogNormal, WeightedIndex};
+use std::time::Duration;
+
+use crate::{NetworkGenerator, PaymentGenerator, SimulationError};
+
+const HOURS_PER_MONTH: u64 = 30 * 24;
+const SECONDS_PER_MONTH: u64 = HOURS_PER_MONTH * 60 * 60;
+
+/// The number of times that our network view will try to pick a node that is *not* the source node provided by the
+/// caller.
+const NETWORK_VIEW_RETRIES: u8 = 5;
+
+/// NetworkGraphView maintains a view of the network graph that can be used to pick nodes by their deployed liquidity
+/// and track node capacity within the network. Tracking nodes in the network is memory-expensive, so we use a single
+/// tracker for the whole network (in an unbounded environment, we'd make one _per_ node generating random activity,
+/// which has a view of the full network except for itself).
+pub struct NetworkGraphView {
+    node_picker: WeightedIndex<u64>,
+    nodes: Vec<(PublicKey, u64)>,
+}
+
+impl NetworkGraphView {
+    pub fn new(node_capacities: HashMap<PublicKey, u64>) -> Result<Self, SimulationError> {
+        // To create a weighted index we're going to need a vector of nodes that we index and weights that are set
+        // by their deployed capacity. To efficiently store our view of nodes capacity, we're also going to store
+        // capacity along with the node pubkey because we query the two at the same time.
+        let nodes = node_capacities.iter().map(|(k, v)| (*k, *v)).collect();
+
+        let node_picker = WeightedIndex::new(node_capacities.into_values().collect::<Vec<u64>>())
+            .map_err(|e| SimulationError::RandomActivityError(e.to_string()))?;
+
+        Ok(NetworkGraphView { node_picker, nodes })
+    }
+}
+
+impl NetworkGenerator for NetworkGraphView {
+    /// Randomly samples the network for a node, weighted by capacity.  Using a single graph view means that it's
+    /// possible for a source node to select itself. This is masked by allowing muliple retries, though it is possible
+    /// that it may still pick the source node in low node-count networks where the source node has proportionally
+    /// more capital deployed than other nodes.
+    fn sample_node_by_capacity(
+        &self,
+        source: PublicKey,
+    ) -> Result<(PublicKey, u64), SimulationError> {
+        let mut rng = rand::thread_rng();
+
+        for _ in 0..NETWORK_VIEW_RETRIES {
+            let index = self.node_picker.sample(&mut rng);
+            let destination = self.nodes[index];
+
+            if destination.0 != source {
+                return Ok(destination);
+            }
+        }
+
+        Err(SimulationError::RandomActivityError(format!(
+            "could not pick node that is not source: {source}"
+        )))
+    }
```

Comment:
I still don't think we should be limiting this loop. If this happened during the simulation, it would actually signal a design flaw in the simulator, more than a wrong configuration.

The reason this can happen is because we have a shared `NetworkGraphView` throughout the generator, hence when we sample, we are among the nodes to be sampled. This is OK because we are trying to avoid copying this for all producers/consumers, but it should be completely transparent to the user.

The potential threat with not limiting this is that the simulator could be slow/get stuck trying to pick a destination, but I actually don't think this could be the case:

For this to happen, it would mean that over several random picks, the same node (the source) is being picked over and over again, however, this would mean that its `node_capacity` is subsequently higher than any other node in the network. The only case in which that can happen is if the network is set up as a star, where the node in the center is connected to every other node in the network, and the other nodes are not connected between them. Otherwise, a node with high capacity will imply other nodes with substantial capacity to also be present and hence will invalidate the original assumption. 

Even in the case of the star network, with a node having 90% of the network's capacity, the odds of subsequently picking it over any other node in the network is `0.9^n` where `n` is the number of iterations. That becomes smaller than `1/2` after 7 iterations.

Reply:
WDYT about including this check as a sanity check that we can use to kill a buggy simulation? Rather than possibly spinning forever.

Eg, allow 20 retries and kill the simulation if we can't pick a non-source node?

Reply:
Discussed offline: we'll only be unable to pick a source node that isn't us with _very, very_ low probability or we have a bug. Decision it to loop and log a warning (with a request to open a bug report).